refactor!: Update config, passwords, use master image

bmtcril opened the pull request:
Mostly cleanup:

- Create real encrypted passwords for the ralph and LMS users. I'm not 100% sure the LMS user belongs here, but Ralph will need one with or without OARS to work! Plus I don't want to spread that ugly hashing any more than I need to. :)
- Use the Ralph master image, we'll likely want to pin this once the latest changes are released
- Make Clickhouse the default (since it's the only one we currently configure). Issue: https://github.com/openedx/tutor-contrib-ralph/issues/3
- Remove currently unused config

bmtcril committed:
refactor!: Update config, passwords, use master image

bmtcril committed:
fix: Use the HTTP port for Clickhouse, fix creds path

pomegranited committed:
refactor!: use OARS variable names for clickhouse LRS user

pomegranited committed:
docs: updates README to mention plugins Ralph relies on

pomegranited commented on tutorralph/plugin.py:
```diff
@@ -149,29 +173,29 @@
 ralph:
-    #image: fundocker/ralph:master
-    image: ralph:development
+    image: docker.io/fundocker/ralph:master
```
To do: pin this to a (configurable) release

pomegranited replied:
I made the docker image and tag configurable with [f7d2465](https://github.com/openedx/tutor-contrib-ralph/pull/4/commits/f7d246544d034496ca689d546493279b641dc543).

bmtcril replied:
Thank you!

pomegranited commented on tutorralph/plugin.py:
```diff
@@ -149,29 +173,29 @@
+      - "clickhouse"
     volumes:
-      - /Users/brianmesick/Dev/forks/ralph:/app
+      - ../../env/plugins/ralph/apps/config/.ralph:/app/.ralph
```
When I first spun all of these plugins up, I was getting connection errors from Ralph about Clickhouse, I think because the Clickhouse init scripts hadn't run yet to create the users?

But I restarted Ralph, and it connected fine, and now I can hit endpoints like this (with the Ralph Admin or LMS credentials): http://localhost:8100/xAPI/statements/

However, when I do that, I only get back: `{"statements":[]}`, so I can't confirm that Ralph can see the xapi data that's loaded into Clickhouse.

How do I verify that Ralph can see this data?

bmtcril replied:
There does seem to be some kind of issue with Ralph spinning up before Clickhouse in spite of the depends_on, I'll look into that today. If you're seeing the `statements` response it's because the database is empty. My install wasn't loading the test xapi data by default, but I'll turn that on today and make sure it's all working.

bmtcril replied:
Ok, I've found both problems. 

1. I'll set up a PR for a Clickhouse healthcheck so Ralph knows when to start
2. Ralph reads from the raw table, the demo data is being written to the parsed table so the data will show up in Superset, but not Ralph. I'll take a look at making the xapi loader write to the raw table instead and let the materialized view handle it.

bmtcril replied:
Ralph reporting no statements on demo data is fixed here: https://github.com/openedx/tutor-contrib-oars/pull/9

pomegranited replied:
I can confirm that this works eventually, but it didn't work from a fresh install :( Please see https://github.com/openedx/tutor-contrib-oars/pull/9#discussion_r1134749884

bmtcril committed:
Merge branch 'main' into bmtcril/use_master_image

bmtcril committed:
fix: Fix ClickHouse dependency

We need to make sure ClickHouse is running, not just starting.

bmtcril commented:
@pomegranited this is ready for re-review, but now depends on https://github.com/openedx/tutor-contrib-clickhouse/pull/6 merging first

pomegranited commented on tutorralph/plugin.py:
```diff
@@ -149,29 +173,30 @@
     depends_on:
-      - clickhouse
+      clickhouse:
+        condition: service_healthy
```
This is fantastic, thank you for working out this fix @bmtcril !
However using this `depends_on` syntax means that we need to be using `docker-compose>=1.27` ([ref](https://stackoverflow.com/a/66485341)), which is working totally fine, it's just a slight step up from [Tutor's requirement of >1.22](https://docs.tutor.overhang.io/install.html#requirements). I noticed it because I was running `1.25`, and it errorred here, saying that `depends_on` expects an array.

Unfortunately, there doesn't seem to be a way to specify this version in the docker-compose yaml, because docker-compose <2.0 is indicated by "omitting the `version`" ([ref](https://docs.docker.com/compose/compose-file/compose-versioning/#versioning)).

So maybe we should put a note in the Ralph and OARS plugin READMEs?

bmtcril replied:
I have a note to do that, but if this isn't actually solving the problem I may just remove it. I'm punting that decision for a couple of weeks, though. :)

pomegranited committed:
feat: makes ralph's docker image and port configurable

pomegranited committed:
temp: use the CLICKHOUSE_ADMIN_USER

Ralph doesn't restart when authentication fails, so we use clickhouse's
Admin user, which will be exist when Ralph starts up.

pomegranited commented on tutorralph/templates/ralph/apps/config/env:
```diff
@@ -1,78 +1,18 @@
-
-# RALPH_CONVERTER_EDX_XAPI_UUID_NAMESPACE=
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__USERNAME={{CLICKHOUSE_ADMIN_USER}}
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__PASSWORD={{CLICKHOUSE_ADMIN_PASSWORD}}
```
FYI @bmtcril -- I tried a bunch of different things to try to address this, but to get it working, I ended up just using the Clickhouse Admin user here :(

I tried:

* Using the `ralph auth ... --write` command-line tool to add the Ralph LRS user, because Ralph's startup text indicated that it was watching `/app` for changes, so I thought it would restart if the `/app/.ralph/auth.json` file changes. It didn't.
   (Also btw, the ralph command-line tool in the docker image is broken, so I had to write my own.)
* Setting docker-compose `healthcheck` and `restart` statements for the ralph service based on successfully hitting the /xAPI/statements endpoint -- I couldn't get the ralph container to restart on failure.
* Creating a second container called `ralph_clickhouse` that `ralph` `depends_on`, which tried checking the clickhouse connection using the CLICKHOUSE_LMS user to indicate its success -- the best I could do was mark the `ralph` container as "unhealthy", but couldn't get anything to restart.

I still think the best actual option is to modify the docker deployment for Ralph to run it as a service that restarts itself until it's running properly, but since I couldn't figure out how to do it with the docker-compose container, that was out of scope here.

bmtcril replied:
Yeah I think the fact that Ralph wedges instead of stopping when it hits that error is problem 1, but I'm reconsidering having the services own their users instead of OARS to prevent these tangles going forward. Presumably even if they weren't being used in an OARS context they would need the same types of users. Still pondering on it, though.

bmtcril committed:
refactor: Make host and protocol configurable

pomegranited reviewed (approved):
:+1: Still some polishing to do, but it's working well enough to merge! Thank you @bmtcril !

- [x] I tested this by spinning up a fresh tutor dev stack with this and the latest code from the other OARS plugins. I watched events get piped from LMS to event routing backends to Ralph to Clickhouse and ultimately to Superset Dashboard! Worked a treat.
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A, backend service
- [x] Includes documentation
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)
